File: src/main/java/org/cubexmc/metro/gui/controller/LineListController.java
package org.cubexmc.metro.gui.controller;

import java.util.List;
import java.util.Map;

import org.bukkit.entity.Player;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.gui.GuiHolder;
import org.cubexmc.metro.gui.GuiSlots;
import org.cubexmc.metro.model.Line;
import org.cubexmc.metro.util.OwnershipUtil;

public final class LineListController {
    private final Metro plugin;

    public LineListController(Metro plugin) {
        this.plugin = plugin;
    }

    public void handleLineListClick(Player player, GuiHolder holder, int slot, boolean isRightClick) {
        int page = holder.getData("page", 0);
        boolean showOnlyMine = holder.getData("showOnlyMine", false);
        int totalPages = holder.getData("totalPages", 1);

        switch (slot) {
            case GuiSlots.SLOT_PREV_PAGE -> {
                if (page > 0) {
                    plugin.getGuiManager().openLineList(player, page - 1, showOnlyMine, holder.getPreviousView());
                }
                return;
            }
            case GuiSlots.SLOT_NEXT_PAGE -> {
                if (page < totalPages - 1) {
                    plugin.getGuiManager().openLineList(player, page + 1, showOnlyMine, holder.getPreviousView());
                }
                return;
            }
            case GuiSlots.SLOT_FILTER -> {
                plugin.getGuiManager().openLineList(player, 0, !showOnlyMine, holder.getPreviousView());
                return;
            }
            case GuiSlots.SLOT_BACK -> {
                plugin.getGuiManager().openPreviousView(player, holder,
                        () -> plugin.getGuiManager().openMainMenu(player));
                return;
            }
            default -> {
            }
        }

        if (slot >= GuiSlots.ITEMS_PER_PAGE) {
            return;
        }

        List<String> lineNames = holder.getData("lineNames");
        Map<String, List<Line>> groupedLines = holder.getData("groupedLines");
        if (lineNames == null || groupedLines == null) {
            return;
        }

        int index = page * GuiSlots.ITEMS_PER_PAGE + slot;
        if (index < 0 || index >= lineNames.size()) {
            return;
        }

        String name = lineNames.get(index);
        List<Line> variants = groupedLines.get(name);
        if (variants == null || variants.isEmpty()) {
            return;
        }

        if (variants.size() > 1) {
            plugin.getGuiManager().openLineVariants(player, name, 0, holder.snapshot());
            return;
        }

        openLine(player, holder, variants.get(0), isRightClick);
    }

    public void handleLineVariantsClick(Player player, GuiHolder holder, int slot, boolean isRightClick) {
        int page = holder.getData("page", 0);
        String lineName = holder.getData("lineName");
        int totalPages = holder.getData("totalPages", 1);

        switch (slot) {
            case GuiSlots.SLOT_PREV_PAGE -> {
                if (page > 0) {
                    plugin.getGuiManager().openLineVariants(player, lineName, page - 1, holder.getPreviousView());
                }
                return;
            }
            case GuiSlots.SLOT_NEXT_PAGE -> {
                if (page < totalPages - 1) {
                    plugin.getGuiManager().openLineVariants(player, lineName, page + 1, holder.getPreviousView());
                }
                return;
            }
            case GuiSlots.SLOT_BACK -> {
                plugin.getGuiManager().openPreviousView(player, holder,
                        () -> plugin.getGuiManager().openLineList(player, 0, false));
                return;
            }
            default -> {
            }
        }

        if (slot >= GuiSlots.ITEMS_PER_PAGE) {
            return;
        }

        List<Line> lines = holder.getData("lines");
        if (lines == null) {
            return;
        }

        int index = page * GuiSlots.ITEMS_PER_PAGE + slot;
        if (index < 0 || index >= lines.size()) {
            return;
        }

        openLine(player, holder, lines.get(index), isRightClick);
    }

    private void openLine(Player player, GuiHolder holder, Line line, boolean isRightClick) {
        if (isRightClick && OwnershipUtil.canManageLine(player, line)) {
            plugin.getGuiManager().openLineSettings(player, line.getId(), holder.snapshot());
        } else {
            plugin.getGuiManager().openLineDetail(player, line.getId(), 0, holder.snapshot());
        }
    }
}

File: src/main/java/org/cubexmc/metro/gui/controller/LineListController.kt
package org.cubexmc.metro.gui.controller

import org.bukkit.entity.Player
import org.cubexmc.metro.Metro
import org.cubexmc.metro.gui.GuiHolder
import org.cubexmc.metro.gui.GuiSlots
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.util.OwnershipUtil

class LineListController(private val plugin: Metro) {

    fun handleLineListClick(player: Player, holder: GuiHolder, slot: Int, isRightClick: Boolean) {
        val page = holder.getData("page", 0)
        val showOnlyMine = holder.getData("showOnlyMine", false)
        val totalPages = holder.getData("totalPages", 1)

        when (slot) {
            GuiSlots.SLOT_PREV_PAGE -> {
                if (page > 0) {
                    plugin.guiManager.openLineList(player, page - 1, showOnlyMine, holder.getPreviousView())
                }
                return
            }

            GuiSlots.SLOT_NEXT_PAGE -> {
                if (page < totalPages - 1) {
                    plugin.guiManager.openLineList(player, page + 1, showOnlyMine, holder.getPreviousView())
                }
                return
            }

            GuiSlots.SLOT_FILTER -> {
                plugin.guiManager.openLineList(player, 0, !showOnlyMine, holder.getPreviousView())
                return
            }

            GuiSlots.SLOT_BACK -> {
                plugin.guiManager.openPreviousView(player, holder) { plugin.guiManager.openMainMenu(player) }
                return
            }

            else -> Unit
        }

        if (slot >= GuiSlots.ITEMS_PER_PAGE) {
            return
        }

        val lineNames = holder.getData<List<String>>("lineNames") ?: return
        val groupedLines = holder.getData<Map<String, List<Line>>>("groupedLines") ?: return

        val index = page * GuiSlots.ITEMS_PER_PAGE + slot
        if (index < 0 || index >= lineNames.size) {
            return
        }

        val name = lineNames[index]
        val variants = groupedLines[name]
        if (variants.isNullOrEmpty()) {
            return
        }

        if (variants.size > 1) {
            plugin.guiManager.openLineVariants(player, name, 0, holder.snapshot())
            return
        }

        openLine(player, holder, variants[0], isRightClick)
    }

    fun handleLineVariantsClick(player: Player, holder: GuiHolder, slot: Int, isRightClick: Boolean) {
        val page = holder.getData("page", 0)
        val lineName = holder.getData<String>("lineName")
        val totalPages = holder.getData("totalPages", 1)

        when (slot) {
            GuiSlots.SLOT_PREV_PAGE -> {
                if (page > 0) {
                    plugin.guiManager.openLineVariants(player, lineName, page - 1, holder.getPreviousView())
                }
                return
            }

            GuiSlots.SLOT_NEXT_PAGE -> {
                if (page < totalPages - 1) {
                    plugin.guiManager.openLineVariants(player, lineName, page + 1, holder.getPreviousView())
                }
                return
            }

            GuiSlots.SLOT_BACK -> {
                plugin.guiManager.openPreviousView(player, holder) {
                    plugin.guiManager.openLineList(player, 0, false)
                }
                return
            }

            else -> Unit
        }

        if (slot >= GuiSlots.ITEMS_PER_PAGE) {
            return
        }

        val lines = holder.getData<List<Line>>("lines") ?: return

        val index = page * GuiSlots.ITEMS_PER_PAGE + slot
        if (index < 0 || index >= lines.size) {
            return
        }

        openLine(player, holder, lines[index], isRightClick)
    }

    private fun openLine(player: Player, holder: GuiHolder, line: Line, isRightClick: Boolean) {
        if (isRightClick && OwnershipUtil.canManageLine(player, line)) {
            plugin.guiManager.openLineSettings(player, line.id, holder.snapshot())
        } else {
            plugin.guiManager.openLineDetail(player, line.id, 0, holder.snapshot())
        }
    }
}
